libpromises/rlist.c

```diff
@@ -325,18 +325,50 @@ Rlist *RlistAppendRval(Rlist **start, Rval rval)
     return rp;
 }
 
-Rval RvalNew(const void *item, RvalType type)
+Rval RvalNewRewriter(const void *item, RvalType type, JsonElement *map)
 {
+    int max_size = 10*CF_BUFSIZE+1;
+    char buffer_from[max_size];
+    char buffer_to[max_size];
+
     switch (type)
     {
     case RVAL_TYPE_SCALAR:
-        return (Rval) { xstrdup(item), RVAL_TYPE_SCALAR };
+        if (NULL != map && JsonLength(map) > 0)
+        {
+            // BufferExpandStr and BufferExpandRegexes functions would
+            // be great instead of this CF_BUFSIZE ugliness.  ExpandScalar
+            Buffer *format = BufferNew();
+            strncpy(buffer_from, item, max_size);
+
+            JsonIterator iter = JsonIteratorInit(map);
+            const char *key;
+
+            while ((key = JsonIteratorNextKey(&iter)))
+            {
+                BufferPrintf(format, "$(%s)", key);
+                ReplaceStr(buffer_from, buffer_to, max_size, BufferData(format), JsonObjectGetAsString(map, key));
+                strncpy(buffer_from, buffer_to, max_size);
+                BufferPrintf(format, "${%s}", key);
+                ReplaceStr(buffer_from, buffer_to, max_size, BufferData(format), JsonObjectGetAsString(map, key));
+            }
```

Comment:
I'd prefer to avoid evaluation here.  If you feel otherwise, tell me why.  Ideally show a test case because I don't see a common need for this kind of replacement.  I added a 3-time iteration to the rewriting loop but hope I can avoid the rest.